### ISSUE TITLE
Migrate C to Rust with Box + Fortran 2018

### ISSUE THREAD

**Issue by baiango** [baiango/C-GUI#3]
# Why Rust
1. C and C++ is a subtle form of Java. "Write once, debug everywhere" with its compiler and build system.

What worked on this compiler doesn't mean it will work on another compiler.
## Introduction
Today, we are going to take this number `0x8F4A2B5C1D3E7A9B` and count numbers of 1 in binary.
You can put it in your [search engine](https://duckduckgo.com/?q=0x8F4A2B5C1D3E7A9B+to+binary&ia=answer) to convert it.
I got `1000111101001010001010110101110000011101001111100111101010011011` as the number.
Let's make this number easier to count by splitting it into 8 lines.
```
10001111 # 5 1s
01001010 # 3
00101011 # 4
01011100 # 4
00011101 # 4
00111110 # 5
01111010 # 4
10011011 # 5
```
Now [add](https://duckduckgo.com/?q=5%2B3%2B4%2B4%2B4%2B5%2B4%2B5&ia=calculator) the number together with `5+3+4+4+4+5+4+5`. The result is 34. Let's do it in code now.
## The code `in C`
```
#include <stdint.h>


int32_t count_1s_in_binary(uint64_t x) {
    uint32_t count = 0;
    for (; x; count++) {
        x &= x - 1; // Clear the smallest bit in 'x'
    }
    return count;
}
```
Now we can check out each compiler's assembly output to see which one has a problem figuring out what we are trying to do.
### [GCC: -O2 -mavx2](https://godbolt.org/#g:!((g:!((g:!((h:codeEditor,i:(filename:'1',fontScale:14,fontUsePx:'0',j:1,lang:___c,selection:(endColumn:2,endLineNumber:10,positionColumn:2,positionLineNumber:10,selectionStartColumn:2,selectionStartLineNumber:10,startColumn:2,startLineNumber:10),source:'%23include+%3Cstdint.h%3E%0A%0A%0Aint32_t+count_1s_in_binary(uint64_t+x)+%7B%0A++++uint32_t+count+%3D+0%3B%0A++++for+(%3B+x%3B+count%2B%2B)+%7B%0A++++++++x+%26%3D+x+-+1%3B+//+Clear+the+smallest+bit+in+!'x!'%0A++++%7D%0A++++return+count%3B%0A%7D'),l:'5',n:'0',o:'C+source+%231',t:'0')),k:33.333333333333336,l:'4',n:'0',o:'',s:0,t:'0'),(g:!((h:compiler,i:(compiler:cg132,deviceViewOpen:'1',filters:(b:'0',binary:'1',binaryObject:'1',commentOnly:'0',debugCalls:'1',demangle:'0',directives:'0',execute:'1',intel:'0',libraryCode:'0',trim:'1'),flagsViewOpen:'1',fontScale:14,fontUsePx:'0',j:1,lang:___c,libs:!(),options:'-O2+-mavx2',overrides:!(),selection:(endColumn:20,endLineNumber:1,positionColumn:1,positionLineNumber:1,selectionStartColumn:20,selectionStartLineNumber:1,startColumn:1,startLineNumber:1),source:1),l:'5',n:'0',o:'+x86-64+gcc+13.2+(Editor+%231)',t:'0')),k:33.333333333333336,l:'4',n:'0',o:'',s:0,t:'0'),(g:!((h:output,i:(compilerName:'x86-64+gcc+13.2',editorid:1,fontScale:14,fontUsePx:'0',j:1,wrap:'1'),l:'5',n:'0',o:'Output+of+x86-64+gcc+13.2+(Compiler+%231)',t:'0')),k:33.33333333333333,l:'4',n:'0',o:'',s:0,t:'0')),l:'2',n:'0',o:'',t:'0')),version:4)
```
count_1s_in_binary:
        xor     eax, eax ; XOR itself to set it to 0, it doesn't do anything??? It doesn't affect the value in rax
        popcnt  rax, rdi ; Count 1s in binary
        ret
```
### [Clang: -O2 -mavx2](https://godbolt.org/#g:!((g:!((g:!((h:codeEditor,i:(filename:'1',fontScale:14,fontUsePx:'0',j:1,lang:___c,selection:(endColumn:2,endLineNumber:10,positionColumn:2,positionLineNumber:10,selectionStartColumn:2,selectionStartLineNumber:10,startColumn:2,startLineNumber:10),source:'%23include+%3Cstdint.h%3E%0A%0A%0Aint32_t+count_1s_in_binary(uint64_t+x)+%7B%0A++++uint32_t+count+%3D+0%3B%0A++++for+(%3B+x%3B+count%2B%2B)+%7B%0A++++++++x+%26%3D+x+-+1%3B+//+Clear+the+smallest+bit+in+!'x!'%0A++++%7D%0A++++return+count%3B%0A%7D'),l:'5',n:'0',o:'C+source+%231',t:'0')),k:33.333333333333336,l:'4',n:'0',o:'',s:0,t:'0'),(g:!((h:compiler,i:(compiler:cclang1600,deviceViewOpen:'1',filters:(b:'0',binary:'1',binaryObject:'1',commentOnly:'0',debugCalls:'1',demangle:'0',directives:'0',execute:'1',intel:'0',libraryCode:'0',trim:'1'),flagsViewOpen:'1',fontScale:14,fontUsePx:'0',j:1,lang:___c,libs:!(),options:'-O2+-mavx2',overrides:!(),selection:(endColumn:1,endLineNumber:1,positionColumn:1,positionLineNumber:1,selectionStartColumn:1,selectionStartLineNumber:1,startColumn:1,startLineNumber:1),source:1),l:'5',n:'0',o:'+x86-64+clang+16.0.0+(Editor+%231)',t:'0')),k:33.333333333333336,l:'4',n:'0',o:'',s:0,t:'0'),(g:!((h:output,i:(compilerName:'x86-64+gcc+13.2',editorid:1,fontScale:14,fontUsePx:'0',j:1,wrap:'1'),l:'5',n:'0',o:'Output+of+x86-64+gcc+13.2+(Compiler+%231)',t:'0')),k:33.33333333333333,l:'4',n:'0',o:'',s:0,t:'0')),l:'2',n:'0',o:'',t:'0')),version:4)
```
count_1s_in_binary:                     # @count_1s_in_binary
        popcnt  rax, rdi ; Count 1s in binary
        ret
```
### [Icx: -O2 -mavx2](https://godbolt.org/#g:!((g:!((g:!((h:codeEditor,i:(filename:'1',fontScale:14,fontUsePx:'0',j:1,lang:___c,selection:(endColumn:2,endLineNumber:10,positionColumn:2,positionLineNumber:10,selectionStartColumn:2,selectionStartLineNumber:10,startColumn:2,startLineNumber:10),source:'%23include+%3Cstdint.h%3E%0A%0A%0Aint32_t+count_1s_in_binary(uint64_t+x)+%7B%0A++++uint32_t+count+%3D+0%3B%0A++++for+(%3B+x%3B+count%2B%2B)+%7B%0A++++++++x+%26%3D+x+-+1%3B+//+Clear+the+smallest+bit+in+!'x!'%0A++++%7D%0A++++return+count%3B%0A%7D'),l:'5',n:'0',o:'C+source+%231',t:'0')),k:33.333333333333336,l:'4',n:'0',o:'',s:0,t:'0'),(g:!((h:compiler,i:(compiler:cicxlatest,deviceViewOpen:'1',filters:(b:'0',binary:'1',binaryObject:'1',commentOnly:'0',debugCalls:'1',demangle:'0',directives:'0',execute:'1',intel:'0',libraryCode:'0',trim:'1'),flagsViewOpen:'1',fontScale:14,fontUsePx:'0',j:1,lang:___c,libs:!(),options:'-O2+-mavx2',overrides:!(),selection:(endColumn:12,endLineNumber:3,positionColumn:12,positionLineNumber:3,selectionStartColumn:12,selectionStartLineNumber:3,startColumn:12,startLineNumber:3),source:1),l:'5',n:'0',o:'+x86-64+icx+(latest)+(Editor+%231)',t:'0')),k:33.333333333333336,l:'4',n:'0',o:'',s:0,t:'0'),(g:!((h:output,i:(compilerName:'x86-64+gcc+13.2',editorid:1,fontScale:14,fontUsePx:'0',j:1,wrap:'1'),l:'5',n:'0',o:'Output+of+x86-64+icx+(latest)+(Compiler+%231)',t:'0')),k:33.33333333333333,l:'4',n:'0',o:'',s:0,t:'0')),l:'2',n:'0',o:'',t:'0')),version:4)
```
count_1s_in_binary:                     # 
        popcnt  rax, rdi ; Count 1s in binary
        ret
```
### [MSVC: -O2 -arch:AVX2](https://godbolt.org/#g:!((g:!((g:!((h:codeEditor,i:(filename:'1',fontScale:14,fontUsePx:'0',j:1,lang:___c,selection:(endColumn:2,endLineNumber:10,positionColumn:2,positionLineNumber:10,selectionStartColumn:2,selectionStartLineNumber:10,startColumn:2,startLineNumber:10),source:'%23include+%3Cstdint.h%3E%0A%0A%0Aint32_t+count_1s_in_binary(uint64_t+x)+%7B%0A++++uint32_t+count+%3D+0%3B%0A++++for+(%3B+x%3B+count%2B%2B)+%7B%0A++++++++x+%26%3D+x+-+1%3B+//+Clear+the+smallest+bit+in+!'x!'%0A++++%7D%0A++++return+count%3B%0A%7D'),l:'5',n:'0',o:'C+source+%231',t:'0')),k:33.333333333333336,l:'4',n:'0',o:'',s:0,t:'0'),(g:!((h:compiler,i:(compiler:vc_v19_latest_x64,deviceViewOpen:'1',filters:(b:'0',binary:'1',binaryObject:'1',commentOnly:'0',debugCalls:'1',demangle:'0',directives:'0',execute:'1',intel:'0',libraryCode:'0',trim:'1'),flagsViewOpen:'1',fontScale:14,fontUsePx:'0',j:1,lang:___c,libs:!(),options:'-O2+-arch:AVX2',overrides:!(),selection:(endColumn:24,endLineNumber:14,positionColumn:24,positionLineNumber:14,selectionStartColumn:24,selectionStartLineNumber:14,startColumn:24,startLineNumber:14),source:1),l:'5',n:'0',o:'+x64+msvc+v19.latest+(Editor+%231)',t:'0')),k:33.333333333333336,l:'4',n:'0',o:'',s:0,t:'0'),(g:!((h:output,i:(compilerName:'x86-64+gcc+13.2',editorid:1,fontScale:14,fontUsePx:'0',j:1,wrap:'1'),l:'5',n:'0',o:'Output+of+x64+msvc+v19.latest+(Compiler+%231)',t:'0')),k:33.33333333333333,l:'4',n:'0',o:'',s:0,t:'0')),l:'2',n:'0',o:'',t:'0')),version:4)
```
x$ = 8
count_1s_in_binary PROC                       ; COMDAT
        xor     eax, eax ; XOR itself(the return register) to set it to 0
        test    rcx, rcx ; If rcx(the input register) is 0
        je      SHORT $LN10@count_1s_i ; If rcx is 0, jump to return.
        npad    9 ; Doesn't do anything
$LL4@count_1s_i:
        blsr    rcx, rcx ; Clear the smallest bit in 'rcx' (binary shift right)
        inc     eax ; Same as eax++
        test    rcx, rcx ; If rcx is 0
        jne     SHORT $LL4@count_1s_i ; Repeat the loop again.
$LN10@count_1s_i:
        ret     0 ; Return eax
count_1s_in_binary ENDP
```
Only MSVC failed to convert the code into 1 instruction. I think you should only use MSVC when you are only using Windows. The issue with MSVC is that it's not the best at optimizing your code.

The only way to solve this problem is to know your ASM instructions like `popcnt`, and `vmulps` and then you can start to use MSVC compiler-specific functions like `__popcnt` and `_mm_mul_ps`. But that will make your code messy very fast by adding `#ifdef _MSC_VER` everywhere in the program if you want to support Linux or MacOS. Here is an example.
```
#include <stdint.h>

#ifdef _MSC_VER
#include <intrin.h>
#endif

int32_t count_1s_in_binary(uint64_t x) {
    uint32_t count = 0;
#ifdef _MSC_VER
    // Use MSVC-specific intrinsic
    count = __popcnt64(x);
#else
    // Use a generic, portable approach
    for (; x; count++) {
        x &= x - 1; // Clear the smallest bit in 'x'
    }
#endif
    return count;
}
```
You probably found this less readable than before.

You can learn new ASM instructions by creating math-intensive calculations like the dot product, cross product, and matrix and then putting them in (ifx)Intel Fortran compiler.
## Use case for popcnt
1. [Avalanche effect](https://en.wikipedia.org/wiki/Avalanche_effect)
2. [Uniform distribution of hash](https://en.wikipedia.org/wiki/SUHA_(computer_science))(SUHA) I did the code [here](https://github.com/baiango/C-GUI/issues/1#issuecomment-1696481948).

### E.g. 2 `Default value for uninitialized float`
Won't do.

3. The lack of auto-free and default initializers makes it a recipe for a debug farm.
# Why Fortran?
1. Faster than C and Rust because of better compilers.
2. Less error prone than C due to less pointer aliasing issues.
### [GFortran: -O2 -march=core-avx2](https://godbolt.org/#g:!((g:!((g:!((h:codeEditor,i:(filename:'1',fontScale:14,fontUsePx:'0',j:1,lang:fortran,selection:(endColumn:1,endLineNumber:18,positionColumn:1,positionLineNumber:18,selectionStartColumn:1,selectionStartLineNumber:1,startColumn:1,startLineNumber:1),source:'module+VecModule%0A++++implicit+none%0A++++type+::+Vec3f%0A++++++++real+::+x,+y,+z%0A++++end+type+Vec3f%0Aend+module+VecModule%0A%0Asubroutine+cgui_cross(vec,+a,+b)%0A++++use+VecModule%0A++++implicit+none%0A++++type(Vec3f),+intent(out)+::+vec%0A++++type(Vec3f),+intent(in)+::+a,+b%0A%0A++++vec%25x+%3D+a%25y+*+b%25z+-+a%25z+*+b%25y%0A++++vec%25y+%3D+a%25z+*+b%25x+-+a%25x+*+b%25z%0A++++vec%25z+%3D+a%25x+*+b%25y+-+a%25y+*+b%25x%0Aend+subroutine+cgui_cross%0A'),l:'5',n:'0',o:'Fortran+source+%231',t:'0')),k:33.333333333333336,l:'4',n:'0',o:'',s:0,t:'0'),(g:!((h:compiler,i:(compiler:gfortran132,deviceViewOpen:'1',filters:(b:'0',binary:'1',binaryObject:'1',commentOnly:'0',debugCalls:'1',demangle:'0',directives:'0',execute:'0',intel:'0',libraryCode:'0',trim:'1'),flagsViewOpen:'1',fontScale:14,fontUsePx:'0',j:1,lang:fortran,libs:!(),options:'-O2+-march%3Dcore-avx2',overrides:!(),selection:(endColumn:1,endLineNumber:1,positionColumn:1,positionLineNumber:1,selectionStartColumn:1,selectionStartLineNumber:1,startColumn:1,startLineNumber:1),source:1),l:'5',n:'0',o:'+x86-64+gfortran+13.2+(Editor+%231)',t:'0')),k:33.333333333333336,l:'4',n:'0',o:'',s:0,t:'0'),(g:!((h:output,i:(compilerName:'x86-64+gcc+13.2',editorid:1,fontScale:14,fontUsePx:'0',j:1,wrap:'1'),l:'5',n:'0',o:'Output+of+x86-64+gfortran+13.2+(Compiler+%231)',t:'0')),k:33.33333333333333,l:'4',n:'0',o:'',s:0,t:'0')),l:'2',n:'0',o:'',t:'0')),version:4)
Note: Variables that start with `i,j,k,l,m,n` will be specified as integers. So you use `implicit none` to disable this feature.
```
module VecModule
    implicit none
    type :: Vec3f
        real :: x, y, z
    end type Vec3f
end module VecModule

subroutine cgui_cross(vec, a, b)
    use VecModule
    implicit none
    type(Vec3f), intent(out) :: vec
    type(Vec3f), intent(in) :: a, b

    vec%x = a%y * b%z - a%z * b%y
    vec%y = a%z * b%x - a%x * b%z
    vec%z = a%x * b%y - a%y * b%x
end subroutine cgui_cross
```
```
cgui_cross_:
        vmovss  xmm1, DWORD PTR [rsi]
        vmovss  xmm7, DWORD PTR [rsi+8]
        vmovq   xmm2, QWORD PTR [rdx+4]
        vmovss  xmm5, DWORD PTR [rdx]
        vunpcklps       xmm4, xmm7, xmm1
        vmovss  xmm6, DWORD PTR [rdx+8]
        vmovq   xmm0, QWORD PTR [rsi+4]
        vmulps  xmm4, xmm4, xmm2
        vunpcklps       xmm3, xmm6, xmm5
        vfmsub132ps     xmm3, xmm4, xmm0
        vmulss  xmm0, xmm0, xmm5
        vmovlps QWORD PTR [rdi], xmm3
        vfmsub132ss     xmm1, xmm0, xmm2
        vmovss  DWORD PTR [rdi+8], xmm1
        ret
```
### [C GCC: -O2 -mavx2](https://godbolt.org/#g:!((g:!((g:!((h:codeEditor,i:(filename:'1',fontScale:14,fontUsePx:'0',j:1,lang:___c,selection:(endColumn:1,endLineNumber:11,positionColumn:1,positionLineNumber:11,selectionStartColumn:1,selectionStartLineNumber:1,startColumn:1,startLineNumber:1),source:'struct+Vec3f+%7B+float+x,+y,+z%3B+%7D%3B%0A%0A%0Avoid+cgui_cross(struct+Vec3f*+restrict+vec,+struct+Vec3f*+a,+struct+Vec3f*+b)%0A%7B%09vec-%3Ex+%3D+a-%3Ey+*+b-%3Ez%0A%09++++-+a-%3Ez+*+b-%3Ey%3B%0A%09vec-%3Ey+%3D+a-%3Ez+*+b-%3Ex%0A%09++++-+a-%3Ex+*+b-%3Ez%3B%0A%09vec-%3Ez+%3D+a-%3Ex+*+b-%3Ey%0A%09++++-+a-%3Ey+*+b-%3Ex%3B+%7D%0A'),l:'5',n:'0',o:'C+source+%231',t:'0')),k:33.333333333333336,l:'4',n:'0',o:'',s:0,t:'0'),(g:!((h:compiler,i:(compiler:cg132,deviceViewOpen:'1',filters:(b:'0',binary:'1',binaryObject:'1',commentOnly:'0',debugCalls:'1',demangle:'0',directives:'0',execute:'0',intel:'0',libraryCode:'0',trim:'1'),flagsViewOpen:'1',fontScale:14,fontUsePx:'0',j:1,lang:___c,libs:!(),options:'-O2+-mavx2',overrides:!(),selection:(endColumn:1,endLineNumber:1,positionColumn:1,positionLineNumber:1,selectionStartColumn:1,selectionStartLineNumber:1,startColumn:1,startLineNumber:1),source:1),l:'5',n:'0',o:'+x86-64+gcc+13.2+(Editor+%231)',t:'0')),k:33.333333333333336,l:'4',n:'0',o:'',s:0,t:'0'),(g:!((h:output,i:(compilerName:'x86-64+gcc+13.2',editorid:1,fontScale:14,fontUsePx:'0',j:1,wrap:'1'),l:'5',n:'0',o:'Output+of+x86-64+gcc+13.2+(Compiler+%231)',t:'0')),k:33.33333333333333,l:'4',n:'0',o:'',s:0,t:'0')),l:'2',n:'0',o:'',t:'0')),version:4)
```
struct Vec3f { float x, y, z; };

void cgui_cross(struct Vec3f* restrict vec, struct Vec3f* a, struct Vec3f* b)
{	vec->x = a->y * b->z
	    - a->z * b->y;
	vec->y = a->z * b->x
	    - a->x * b->z;
	vec->z = a->x * b->y
	    - a->y * b->x; }
```
```
cgui_cross:
        vmovss  xmm4, DWORD PTR [rdx]
        vmovss  xmm5, DWORD PTR [rsi]
        vmovss  xmm6, DWORD PTR [rdx+8]
        vmovss  xmm7, DWORD PTR [rsi+8]
        vmovq   xmm1, QWORD PTR [rsi+4]
        vmovq   xmm0, QWORD PTR [rdx+4]
        vunpcklps       xmm2, xmm6, xmm4
        vunpcklps       xmm3, xmm7, xmm5
        vmulps  xmm2, xmm2, xmm1
        vmulps  xmm3, xmm3, xmm0
        vmulss  xmm1, xmm1, xmm4
        vmulss  xmm0, xmm0, xmm5
        vsubps  xmm2, xmm2, xmm3
        vsubss  xmm0, xmm0, xmm1
        vmovlps QWORD PTR [rdi], xmm2
        vmovss  DWORD PTR [rdi+8], xmm0
        ret
```
### [Rustc: -C opt-level=2 -C target-feature=+avx2,+fma](https://godbolt.org/#g:!((g:!((g:!((h:codeEditor,i:(filename:'1',fontScale:14,fontUsePx:'0',j:1,lang:rust,selection:(endColumn:1,endLineNumber:8,positionColumn:1,positionLineNumber:8,selectionStartColumn:1,selectionStartLineNumber:8,startColumn:1,startLineNumber:8),source:'pub+struct+Vec3f+%7B+x:+f32,+y:+f32,+z:+f32+%7D%0A%0Apub+fn+cross(vec:+%26mut+Vec3f,+a:+%26Vec3f,+b:+%26Vec3f)+%7B%0A++++vec.x+%3D+a.y+*+b.z+-+a.z+*+b.y%3B%0A++++vec.y+%3D+a.z+*+b.x+-+a.x+*+b.z%3B%0A++++vec.z+%3D+a.x+*+b.y+-+a.y+*+b.x%3B%0A%7D%0A'),l:'5',n:'0',o:'Rust+source+%231',t:'0')),k:33.333333333333336,l:'4',n:'0',o:'',s:0,t:'0'),(g:!((h:compiler,i:(compiler:r1720,deviceViewOpen:'1',filters:(b:'0',binary:'1',binaryObject:'1',commentOnly:'0',debugCalls:'1',demangle:'0',directives:'0',execute:'1',intel:'0',libraryCode:'0',trim:'1'),flagsViewOpen:'1',fontScale:14,fontUsePx:'0',j:1,lang:rust,libs:!(),options:'-C+opt-level%3D2+-C+target-feature%3D%2Bavx2,%2Bfma',overrides:!(),selection:(endColumn:1,endLineNumber:1,positionColumn:1,positionLineNumber:1,selectionStartColumn:1,selectionStartLineNumber:1,startColumn:1,startLineNumber:1),source:1),l:'5',n:'0',o:'+rustc+1.72.0+(Editor+%231)',t:'0')),k:33.333333333333336,l:'4',n:'0',o:'',s:0,t:'0'),(g:!((h:output,i:(compilerName:'rustc+1.72.0',editorid:1,fontScale:14,fontUsePx:'0',j:1,wrap:'1'),l:'5',n:'0',o:'Output+of+rustc+1.72.0+(Compiler+%231)',t:'0')),k:33.33333333333333,l:'4',n:'0',o:'',s:0,t:'0')),l:'2',n:'0',o:'',t:'0')),version:4)
```
pub struct Vec3f { x: f32, y: f32, z: f32 }

pub fn cross(vec: &mut Vec3f, a: &Vec3f, b: &Vec3f) {
    vec.x = a.y * b.z - a.z * b.y;
    vec.y = a.z * b.x - a.x * b.z;
    vec.z = a.x * b.y - a.y * b.x;
}
```
```
example::cross:
        vmovss  xmm0, dword ptr [rdx + 4]
        vmovss  xmm1, dword ptr [rdx + 8]
        vmovss  xmm2, dword ptr [rsi + 4]
        vmovss  xmm3, dword ptr [rsi + 8]
        vmulss  xmm4, xmm2, xmm1
        vmulss  xmm5, xmm3, xmm0
        vsubss  xmm4, xmm4, xmm5
        vmovss  dword ptr [rdi], xmm4
        vmovss  xmm4, dword ptr [rdx]
        vmulss  xmm3, xmm3, xmm4
        vmovss  xmm5, dword ptr [rsi]
        vmulss  xmm1, xmm1, xmm5
        vsubss  xmm1, xmm3, xmm1
        vmovss  dword ptr [rdi + 4], xmm1
        vmulss  xmm0, xmm0, xmm5
        vmulss  xmm1, xmm2, xmm4
        vsubss  xmm0, xmm0, xmm1
        vmovss  dword ptr [rdi + 8], xmm0
        ret
```
GFortran was able to find a shortcut than C and used `vfmsub132ps` to cut down 2 instructions.
### The Latency [`Credit here`](https://stackoverflow.com/questions/692718/how-many-cpu-cycles-are-needed-for-each-assembly-instruction)
The rules for counting latency are based on this [instruction table](https://www.agner.org/optimize/instruction_tables.pdf) on Intel Coffee Lake(9th gen Intel CPU). Because that is what I am using now. The registers starting from `e`(32-bit) and `r`(64-bit) as in `rbx` will be considered as 64-bit or `r`.
```
add eax, eax
add eax, eax
add eax, eax
add eax, eax # Total latency of 4 cycles for these 4 adds
```
```
add eax, eax
add ebx, ebx
add ecx, ecx
add edx, edx # These 4 instructions might all execute, in parallel in a single cycle
```
Total latency of 4 cycles for these 4 adds, because those are dependent on eax. Left is the output, right are inputs. E.g.
Instruction | output                   | input
------------|--------------------------|-----------------------
add         | eax                      | eax

| Instruction | Operands | μops fused domain | μops unfused domain | μops each port | Latency | Reciprocal throughput | Comments |
|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|
| ADD SUB | r,r/i | 1 | 1 | p0156 | 1 | 0.25 |  |

How do I read it? By dividing 1 by `Reciprocal throughput` and then you get 4. You can do it 4 times in 1 latency. That means 4 independent chained `add` or `sub` will be executed in a single latency.

This might be inaccurate, but I will give it a try.
### [GFortran: -O2 -march=core-avx2](https://godbolt.org/#g:!((g:!((g:!((h:codeEditor,i:(filename:'1',fontScale:14,fontUsePx:'0',j:1,lang:fortran,selection:(endColumn:1,endLineNumber:18,positionColumn:1,positionLineNumber:18,selectionStartColumn:1,selectionStartLineNumber:1,startColumn:1,startLineNumber:1),source:'module+VecModule%0A++++implicit+none%0A++++type+::+Vec3f%0A++++++++real+::+x,+y,+z%0A++++end+type+Vec3f%0Aend+module+VecModule%0A%0Asubroutine+cgui_cross(vec,+a,+b)%0A++++use+VecModule%0A++++implicit+none%0A++++type(Vec3f),+intent(out)+::+vec%0A++++type(Vec3f),+intent(in)+::+a,+b%0A%0A++++vec%25x+%3D+a%25y+*+b%25z+-+a%25z+*+b%25y%0A++++vec%25y+%3D+a%25z+*+b%25x+-+a%25x+*+b%25z%0A++++vec%25z+%3D+a%25x+*+b%25y+-+a%25y+*+b%25x%0Aend+subroutine+cgui_cross%0A'),l:'5',n:'0',o:'Fortran+source+%231',t:'0')),k:33.333333333333336,l:'4',n:'0',o:'',s:0,t:'0'),(g:!((h:compiler,i:(compiler:gfortran132,deviceViewOpen:'1',filters:(b:'0',binary:'1',binaryObject:'1',commentOnly:'0',debugCalls:'1',demangle:'0',directives:'0',execute:'0',intel:'0',libraryCode:'0',trim:'1'),flagsViewOpen:'1',fontScale:14,fontUsePx:'0',j:1,lang:fortran,libs:!(),options:'-O2+-march%3Dcore-avx2',overrides:!(),selection:(endColumn:1,endLineNumber:1,positionColumn:1,positionLineNumber:1,selectionStartColumn:1,selectionStartLineNumber:1,startColumn:1,startLineNumber:1),source:1),l:'5',n:'0',o:'+x86-64+gfortran+13.2+(Editor+%231)',t:'0')),k:33.333333333333336,l:'4',n:'0',o:'',s:0,t:'0'),(g:!((h:output,i:(compilerName:'x86-64+gcc+13.2',editorid:1,fontScale:14,fontUsePx:'0',j:1,wrap:'1'),l:'5',n:'0',o:'Output+of+x86-64+gfortran+13.2+(Compiler+%231)',t:'0')),k:33.33333333333333,l:'4',n:'0',o:'',s:0,t:'0')),l:'2',n:'0',o:'',t:'0')),version:4)
```
cgui_cross_:
        vmovss  xmm1, DWORD PTR [rsi] # Latency 2, Reciprocal Throughput 0.5
        vmovss  xmm7, DWORD PTR [rsi+8]
        # 2 latency
        vmovq   xmm2, QWORD PTR [rdx+4] # L 2, T 1
        # 4 latency
        vmovss  xmm5, DWORD PTR [rdx]
        # 6 latency
        vunpcklps       xmm4, xmm7, xmm1 # L 2, T 2
        # 8 latency
        vmovss  xmm6, DWORD PTR [rdx+8]
        # 10 latency
        vmovq   xmm0, QWORD PTR [rsi+4]
        # 12 latency
        vmulps  xmm4, xmm4, xmm2 # L 4, T 1
        # 16 latency
        vunpcklps       xmm3, xmm6, xmm5
        # 18 latency
        vfmsub132ps     xmm3, xmm4, xmm0 # ?, It's mul + sub. So I guess it's L 5.
        # 23 latency
        vmulss  xmm0, xmm0, xmm5 # L 4, T 0.5
        # 27 latency
        vmovlps QWORD PTR [rdi], xmm3 # L 3, T 1
        # 30 latency
        vfmsub132ss     xmm1, xmm0, xmm2 # ?, It's mul + sub. So I guess it's L 5.
        # 35 latency
        vmovss  DWORD PTR [rdi+8], xmm1
        # 37 latency
        ret
```
There are a lot of different instructions, it will cause stalls. I am not impressed.
### [ifx: -O2 -march=core-avx2](https://godbolt.org/#g:!((g:!((g:!((h:codeEditor,i:(filename:'1',fontScale:14,fontUsePx:'0',j:1,lang:fortran,selection:(endColumn:1,endLineNumber:18,positionColumn:1,positionLineNumber:18,selectionStartColumn:1,selectionStartLineNumber:18,startColumn:1,startLineNumber:18),source:'module+VecModule%0A++++implicit+none%0A++++type+::+Vec3f%0A++++++++real+::+x,+y,+z%0A++++end+type+Vec3f%0Aend+module+VecModule%0A%0Asubroutine+cgui_cross(vec,+a,+b)%0A++++use+VecModule%0A++++implicit+none%0A++++type(Vec3f),+intent(out)+::+vec%0A++++type(Vec3f),+intent(in)+::+a,+b%0A%0A++++vec%25x+%3D+a%25y+*+b%25z+-+a%25z+*+b%25y%0A++++vec%25y+%3D+a%25z+*+b%25x+-+a%25x+*+b%25z%0A++++vec%25z+%3D+a%25x+*+b%25y+-+a%25y+*+b%25x%0Aend+subroutine+cgui_cross%0A'),l:'5',n:'0',o:'Fortran+source+%231',t:'0')),k:33.333333333333336,l:'4',n:'0',o:'',s:0,t:'0'),(g:!((h:compiler,i:(compiler:ifxlatest,deviceViewOpen:'1',filters:(b:'0',binary:'1',binaryObject:'1',commentOnly:'0',debugCalls:'1',demangle:'0',directives:'0',execute:'0',intel:'0',libraryCode:'0',trim:'1'),flagsViewOpen:'1',fontScale:14,fontUsePx:'0',j:1,lang:fortran,libs:!(),options:'-O2+-march%3Dcore-avx2',overrides:!(),selection:(endColumn:12,endLineNumber:20,positionColumn:1,positionLineNumber:4,selectionStartColumn:12,selectionStartLineNumber:20,startColumn:1,startLineNumber:4),source:1),l:'5',n:'0',o:'+x86-64+ifx+(latest)+(Editor+%231)',t:'0')),k:33.333333333333336,l:'4',n:'0',o:'',s:0,t:'0'),(g:!((h:output,i:(compilerName:'x86-64+gcc+13.2',editorid:1,fontScale:14,fontUsePx:'0',j:1,wrap:'1'),l:'5',n:'0',o:'Output+of+x86-64+ifx+(latest)+(Compiler+%231)',t:'0')),k:33.33333333333333,l:'4',n:'0',o:'',s:0,t:'0')),l:'2',n:'0',o:'',t:'0')),version:4)
```
cgui_cross_:
        vmovss  xmm0, dword ptr [rdx + 4] # Latency 2, Reciprocal Throughput 0.5
        vmovss  xmm1, dword ptr [rdx + 8]
        # 2 latency
        vmovss  xmm2, dword ptr [rsi + 4]
        vmovss  xmm3, dword ptr [rsi + 8]
        # 4 latency
        vmulss  xmm4, xmm0, xmm3 # L 4, T 0.5
        # 8 latency
        vfmsub231ss     xmm4, xmm1, xmm2 # ?, It's mul + sub. So I guess is 5
        # 13 latency
        vmovss  dword ptr [rdi], xmm4
        vmovss  xmm4, dword ptr [rdx]
        # 15 latency
        vmovss  xmm5, dword ptr [rsi]
        # 17 latency
        vmulss  xmm1, xmm5, xmm1
        # 21 latency
        vfmsub231ss     xmm1, xmm4, xmm3
        # 26 latency
        vmovss  dword ptr [rdi + 4], xmm1
        # 28 latency
        vmulss  xmm1, xmm4, xmm2
        # 30 latency
        vfmsub231ss     xmm1, xmm5, xmm0
        # 35 latency
        vmovss  dword ptr [rdi + 8], xmm1
        # 37 latency
        ret
```
4 consecutive vmovss(128-bit) instructions? This might run 4 instructions parallelly in future CPU. The CPU with AVX-512 support. It's an impressive future-proof code. For nowadays CPU, it's mostly 2.

Both are 37 latency. But in the future, I think IFX will run faster than GFortran on a CPU with AVX-512 support with 33 latency.

# How
# C functions alternative in Rust
C      | Rust                     | C++
-------|--------------------------|-----------------------
malloc | std::alloc::alloc with Box | new with std::unique_ptr

**Comment by baiango** [comment]
I have concluded that `Box with std::alloc::alloc` in Rust is as fast as `malloc` in C but comes with auto-free.
Look at the total time at the bottom right corner! And the total time.
I remembered that `std::unique_ptr with new` is slower than raw `new` in C++. So I am very skeptical about `Box`.
**Using `Vec` is a very bad idea if you want bare-metal performance.**
# std::alloc::alloc only
```
use std::alloc::{alloc, dealloc, Layout};
use std::ptr;

fn main() {
	let num_elements = 1_000_000_0;
	let element_size = std::mem::size_of::<i32>();
	let total_size = num_elements * element_size;

	// Create a Layout that describes the allocation
	let layout = Layout::from_size_align(total_size, std::mem::align_of::<i32>()).unwrap();

	// Allocate memory using alloc
	let ptr = unsafe { alloc(layout) as *mut i32 };

	if ptr.is_null() {
		println!("Memory allocation failed.");
	}
	// Initialize the memory
	for i in 0..num_elements {
		unsafe {
			ptr::write(ptr.offset(i as isize), (num_elements - i) as i32);
		}
	}
	// Print the contents of the array
	for i in 0..10 {
		unsafe {
			println!("Element at index {}: {}", i, *ptr.offset(i as isize));
		}
	}
	// Deallocate memory when done
	unsafe {
		dealloc(ptr as *mut u8, layout);
	}
}
```
![advixe-gui_oo2pFU4Tw4](https://github.com/baiango/C-GUI/assets/105705580/43ccb2e3-aba2-4fc4-ae78-2cd1523471f6)
# Box with std::alloc::alloc
```
use std::alloc::{alloc, Layout};


pub fn main() {
	const NUM_ELEMENTS: usize = 1_000_000_0;
	let total_size = NUM_ELEMENTS * std::mem::size_of::<i32>();

	// Create a Layout that describes the allocation
	let layout = Layout::from_size_align(total_size, std::mem::align_of::<i32>()).unwrap();

	// Allocate memory using alloc
	let ptr = unsafe { alloc(layout) as *mut [i32; NUM_ELEMENTS] };

	if ptr.is_null() {
		println!("Memory allocation failed.");
		return;
	}

	// Wrap the allocated memory in a Box
	let mut boxed_data = unsafe { Box::from_raw(ptr) };

	// Initialize the memory
	for i in 0..NUM_ELEMENTS {
		boxed_data[i] = (NUM_ELEMENTS - i) as i32;
	}

	// Print the contents of the array
	for i in 0..10 {
		println!("Element at index {}: {}", i, boxed_data[i]);
	}

	// The Box will automatically deallocate memory when it goes out of scope
}
```
![advixe-gui_v1r3ijbU6b](https://github.com/baiango/C-GUI/assets/105705580/6c274bc3-417b-4710-96fc-1d3a0b5e433b)
# Vec
```
pub fn main() {
	let num_elements = 1_000_000_0;

	// Create a Vec containing 1,000,000,0 i32 elements
	let mut vec_data = Vec::with_capacity(num_elements);

	// Initialize the memory
	for i in 0..num_elements {
		vec_data.push((num_elements - i) as i32);
	}

	// Print the contents of the array
	for i in 0..10 {
		println!("Element at index {}: {}", i, vec_data[i]);
	}

	// Memory is automatically deallocated when the Vec goes out of scope
}
```
![advixe-gui_F3csGDz4mV](https://github.com/baiango/C-GUI/assets/105705580/865bc70a-65b2-485b-a5f0-54b6ef42870a)

**Comment by baiango** [comment]
I managed to combine Rust + Fortran together in 441b93c.
You need to use `nm.exe` from GCC to list symbols from Fortran to port it into Rust.
```
nm glmlib.lib

glmath.obj:
0000000000000000 b .bss
0000000000000000 d .data
0000000000000000 N .debug$S
0000000000000000 N .debug$T
0000000000000000 i .drectve
0000000000000000 p .pdata
0000000000000000 t .text
0000000000000000 r .trace
0000000000000000 r .xdata
0000000000000000 a @feat.00
                 U _fltused
0000000000000010 T CGUI_CROSS
0000000000000000 T VECMODULE.
```
# glmath.f90
```
module VecModule
    implicit none
    type :: Vec3f
        real :: x, y, z
    end type Vec3f
end module VecModule

subroutine cgui_cross(vec, a, b)
    use VecModule
    implicit none
    type(Vec3f), intent(out) :: vec
    type(Vec3f), intent(in) :: a, b

    vec%x = a%y * b%z - a%z * b%y
    vec%y = a%z * b%x - a%x * b%z
    vec%z = a%x * b%y - a%y * b%x
end subroutine cgui_cross
```
# main.rs
```
extern {
	// Declare the Fortran subroutine and its arguments
	fn CGUI_CROSS(vec: *mut Vec3f, a: *const Vec3f, b: *const Vec3f);
}

// Define the Vec3f type for Rust
#[repr(C)]
pub struct Vec3f {
	pub x: f32,
	pub y: f32,
	pub z: f32,
}
```

**Comment by baiango** [comment]
This proposal will be rewritten to sound less aggressive and overly negative.

**Comment by baiango** [comment]
The gllib code is fully ported to Fortran. 7532137
But the bindings are not fully made.
- The `/c` options will prevent pdb from being generated.
- The `/Qipo /Qopt-report:1` will corrupt the obj file.
